[mlir][docs] Migrate code examples to nanobind, make Python spelling …

nicholasjng opened the pull request:
…consistent

Since the bindings now use nanobind, I changed the code examples and mentions in the documentation prose to mention nanobind concepts and symbols wherever applicable.

I also made the spelling of "Python" consistent by choosing the uppercase name everywhere that's not an executable name, part of a URL, or directory name.

----------------

Note that I left mentions of `PybindAdaptors.h` in because of https://github.com/llvm/llvm-project/pull/162309.

Are there any thoughts about adding a virtual environment setup guide using [uv](https://docs.astral.sh/uv/)? It has gotten pretty popular, and is much faster than a "vanilla" Python pip install. It can also bootstrap an interpreter not present on the user's machine, for example a free-threaded Python build, with the `-p` flag to the `uv venv` virtual environment creation command.

nicholasjng committed:
[mlir][docs] Migrate code examples to nanobind, make Python spelling consistent

Since the bindings now use nanobind, I changed the code examples and mentions in
the documentation prose to mention nanobind concepts and symbols wherever applicable.

I also made the spelling of "Python" consistent by choosing the uppercase name everywhere
that's not an executable name, part of a URL, or directory name.

jpienaar reviewed (approved):
No objection to adding a uv example. Its not like its being made a requirement and indeed seems rather popular. Python though does seem to grow many tools quickly, so wouldn't want long list of possible ways - but 2 isn't bad where both of them seem in rather wide use.

makslevental reviewed (approved):
lol thanks - this `p`-`P` thing makes me twitch every time I notice it.

nicholasjng commented:
> @nicholasjng let me know if you need this merged for you (not sure if you have write access).

Thanks! I do not, that would be great. You can also hold off until I push the uv bit, though, if you prefer to include it now.

makslevental commented:
> if you prefer to include it now

up to you - no rush on my side

nicholasjng committed:
[mlir][docs] Add note on virtual environment management with uv

Includes a venv creation guide similar to the existing one for `python -m venv`.
Explains all necessary background info on why it works, what it does in the background
(i.e. toolchain downloads), and links to the uv docs for more info so that the reader
has a reference ready when trying it out.

nicholasjng commented:
uv notes were added, happy to hear your thoughts!

nicholasjng commented:
> @nicholasjng LGTM! Should I merge now?

that would be lovely :)
